pallets/merkle/src/utils/hasher.rs

```diff
@@ -74,59 +109,93 @@ impl Setup {
 		Self { hasher, backend }
 	}
 
-	pub fn hash(&self, xl: &ScalarBytes, xr: &ScalarBytes) -> ScalarBytes {
+	pub fn hash<C: Config>(&self, xl: &ScalarBytes, xr: &ScalarBytes) -> Result<ScalarBytes, Error<C>> {
 		match self.backend {
-			Backend::Bulletproofs => match self.hasher {
+			Backend::Bulletproofs(Curve::Curve25519) => match self.hasher {
 				HashFunction::PoseidonDefault | HashFunction::Poseidon(6, 3) => {
 					let sl = Scalar::from_bytes_mod_order(slice_to_bytes_32(xl));
 					let sr = Scalar::from_bytes_mod_order(slice_to_bytes_32(xr));
-					Poseidon_hash_2(sl, sr, &DEFAULT_POSEIDON_HASHER).to_bytes().to_vec()
+					Ok(Poseidon_hash_2(sl, sr, &DEFAULT_POSEIDON_HASHER).to_bytes().to_vec())
 				}
-				_ => panic!("Unsupported hash function"),
+				_ => Err(Error::<C>::Unimplemented),
 			},
-			_ => panic!("Unsupported backend"),
+			Backend::Arkworks(Curve::Bls381, _) => match self.hasher {
+				HashFunction::PoseidonDefault => {
+					let mut bytes = Vec::new();
+					bytes.extend(xl);
+					bytes.extend(xr);
+					let res = PoseidonCRH3::evaluate(&POSEIDON_PARAMETERS, &bytes).unwrap();
+					let bytes_res = to_bytes![res];
+					let bytes = match bytes_res {
+						Ok(bytes) => bytes,
+						Err(_) => return Err(Error::<C>::HashingFailed),
+					};
+					Ok(bytes)
+				}
+				_ => Err(Error::<C>::Unimplemented),
+			},
+			_ => Err(Error::<C>::Unimplemented),
 		}
 	}
 
-	pub fn get_bulletproofs_poseidon(&self) -> &Poseidon {
-		match self.backend {
-			Backend::Bulletproofs => match self.hasher {
-				HashFunction::PoseidonDefault => &DEFAULT_POSEIDON_HASHER,
-				_ => panic!("Hasher is not default poseidon hasher"),
-			},
-			_ => panic!("Backend not bulletproofs"),
+	pub fn get_default_bulletproofs_poseidon<C: Config>(&self) -> Result<&Poseidon, Error<C>> {
+		match (&self.backend, &self.hasher) {
+			(Backend::Bulletproofs(Curve::Curve25519), HashFunction::PoseidonDefault) => Ok(&DEFAULT_POSEIDON_HASHER),
+			_ => Err(Error::<C>::Unimplemented),
 		}
 	}
 
-	pub fn generate_zero_tree(&self, depth: usize) -> (Vec<ScalarBytes>, ScalarBytes) {
-		let zero_tree = match self.hasher {
-			HashFunction::PoseidonDefault => {
-				gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox)
-			}
-			_ => gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox),
-		};
-		(
-			zero_tree[0..depth].iter().map(|x| x.to_vec()).collect(),
-			zero_tree[depth].to_vec(),
-		)
+	pub fn get_default_proving_key<C: Config>(&self) -> Result<&ProvingKey<Bls12_381>, Error<C>> {
```

Comment:
This might not be good practice since proving keys are circuit specific not curve/snark necessarily. Also why are we adding the proving key on-chain? Maybe you're looking for default verifying key instead?

How do you think we will plan to support different trees with the same curve/snark having different verifying keys? Is this something you've specced out or thought of more?

Reply:
This was for testing purposes, but I agree it shouldn't be there.

I think the only way to do this is with storage if we want to add verifying keys on the fly. Lazy static won't work since its not persistent storage, so we could have a `map(Backend, Vec<u8>)` in Merkle storage

Reply:
We could also consider storing a hash of the proving key on-chain. I could see it being useful for a trustless experience. If a website serves you a proving key, you'd be able to ensure it's correct using the on-chain hash identifier.